File: Arrays/Derived_Structures/Heap_Array/Implementation/C++/binary_heap_array.cpp
#include <bits/stdc++.h>
using namespace std;

class BinaryHeap {
	vector<int> arr;
public:
	void push(int x) {
		arr.push_back(x);
		push_heap(arr.begin(), arr.end());
	}
	void pop() {
		pop_heap(arr.begin(), arr.end());
		arr.pop_back();
	}
	int top() { return arr.empty() ? -1 : arr.front(); }
	bool empty() { return arr.empty(); }
};

int main() {
	BinaryHeap h;
	h.push(3); h.push(1); h.push(2);
	cout << h.top() << endl;
	h.pop();
	cout << h.top() << endl;
	cout << (h.empty() ? "Empty" : "Not Empty") << endl;
	return 0;
}

File: Arrays/Derived_Structures/Heap_Array/Implementation/Python/binary_heap_array.py
import heapq

class BinaryHeap:
	def __init__(self):
		self.arr = []
	def push(self, x):
		heapq.heappush(self.arr, x)
	def pop(self):
		if self.arr:
			heapq.heappop(self.arr)
	def top(self):
		return self.arr[0] if self.arr else None
	def empty(self):
		return not self.arr

if __name__ == "__main__":
	h = BinaryHeap()
	h.push(3); h.push(1); h.push(2)
	print(h.top())
	h.pop()
	print(h.top())
	print("Empty" if h.empty() else "Not Empty")
